java/statestore/src/main/java/sleeper/statestore/transactionlog/DynamoDBTransactionLogSnapshotStore.java

```diff
@@ -0,0 +1,137 @@
+        saveSnapshot(SnapshotType.FILES, snapshotPath, transactionNumber);
+    }
+
+    @Override
+    public void savePartitions(String snapshotPath, long transactionNumber) throws DuplicateSnapshotException {
+        saveSnapshot(SnapshotType.PARTITIONS, snapshotPath, transactionNumber);
+    }
+
+    private void saveSnapshot(SnapshotType snapshotType, String snapshotPath, long transactionNumber) throws DuplicateSnapshotException {
+        Instant updateTime = timeSupplier.get();
+        try {
+            dynamo.putItem(new PutItemRequest()
+                    .withTableName(tableName)
+                    .withItem(new DynamoDBRecordBuilder()
+                            .string(TABLE_ID_AND_SNAPSHOT_TYPE, tableAndType(tableName, snapshotType))
+                            .string(TABLE_ID, sleeperTableId)
+                            .string(PATH, snapshotPath)
+                            .number(TRANSACTION_NUMBER, transactionNumber)
+                            .number(UPDATE_TIME, updateTime.toEpochMilli())
+                            .string(SNAPSHOT_TYPE, snapshotType.name())
+                            .build())
+                    .withConditionExpression("attribute_not_exists(#Path)")
+                    .withExpressionAttributeNames(Map.of("#Path", PATH)));
+        } catch (ConditionalCheckFailedException e) {
+            throw new DuplicateSnapshotException(snapshotPath, e);
+        }
+    }
+
+    @Override
+    public List<TransactionLogSnapshot> getFilesSnapshots() {
+        return getSnapshots(SnapshotType.FILES)
+                .collect(Collectors.toList());
+    }
+
+    @Override
+    public List<TransactionLogSnapshot> getPartitionsSnapshots() {
+        return getSnapshots(SnapshotType.PARTITIONS)
+                .collect(Collectors.toList());
+    }
+
+    private Stream<TransactionLogSnapshot> getSnapshots(SnapshotType type) {
+        return streamPagedItems(dynamo, new QueryRequest()
+                .withTableName(tableName)
+                .withConsistentRead(true)
+                .withKeyConditionExpression("#TableIdAndType = :table_and_type")
+                .withExpressionAttributeNames(Map.of("#TableIdAndType", TABLE_ID_AND_SNAPSHOT_TYPE))
+                .withExpressionAttributeValues(new DynamoDBRecordBuilder()
+                        .string(":table_and_type", tableAndType(tableName, type))
+                        .build())
+                .withScanIndexForward(false)
+                .withReturnConsumedCapacity(ReturnConsumedCapacity.TOTAL))
+                .map(this::getSnapshotFromItem);
+    }
+
+    @Override
+    public Optional<TransactionLogSnapshot> getLatestFilesSnapshot() {
+        return getSnapshots(SnapshotType.FILES).findFirst();
+    }
+
+    @Override
+    public Optional<TransactionLogSnapshot> getLatestPartitionsSnapshot() {
+        return getSnapshots(SnapshotType.PARTITIONS).findFirst();
+    }
```

Comment:
This seems a bit inefficient, as it will load a whole page of snapshots, one for files and one for partitions.

Please can we set this up so that there's just one DynamoDB item that needs to be read to find the latest snapshots for a table? It should be possible to query for that directly. If we need a separate table for that, that's fine.